File: sdk/java-sdk/src/main/java/com/flagship/sdk/Evaluator.java
package com.flagship.sdk;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.List;
import java.util.Map;

public class Evaluator {

    public boolean evaluate(FlagConfig flag, String userId, Map<String, String> userAttributes) {
        if (!flag.isEnabled()) return false;
        if (!checkTargetingRules(flag.getTargetingRules(), userAttributes)) return false;
        return bucket(userId, flag.getFlagName(), flag.getSeed()) < flag.getRolloutPercent();
    }

    private boolean checkTargetingRules(Map<String, List<String>> rules, Map<String, String> userAttributes) {
        for (Map.Entry<String, List<String>> entry : rules.entrySet()) {
            String userValue = userAttributes.get(entry.getKey());
            if (userValue == null || !entry.getValue().contains(userValue)) return false;
        }
        return true;
    }

    private int bucket(String userId, String flagName, String seed) {
        try {
            String input = userId + "|" + flagName + "|" + seed;
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(input.getBytes(StandardCharsets.UTF_8));
            long value = 0;
            for (int i = 0; i < 8; i++) {
                value = (value << 8) | (hash[i] & 0xFFL);
            }
            return (int) Long.remainderUnsigned(value, 100);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }
}

File: JS-SDK/src/evaluator.js
function checkTargetingRules(targetingRules, userAttributes) {
  for (const [key, values] of Object.entries(targetingRules)) {
    const userValue = userAttributes[key];
    if (!userValue || !values.includes(userValue)) return false;
  }
  return true;
}

async function bucket(userId, flagName, seed) {
  const input = `${userId}|${flagName}|${seed}`;
  const encoded = new TextEncoder().encode(input);
  const hashBuffer = await globalThis.crypto.subtle.digest('SHA-256', encoded);
  const bytes = new Uint8Array(hashBuffer);
  let value = 0n;
  for (let i = 0; i < 8; i++) value = (value << 8n) | BigInt(bytes[i]);
  return Number(value % 100n);
}

export async function evaluate(flag, userId, userAttributes = {}) {
  if (!flag.enabled) return false;
  if (!checkTargetingRules(flag.targetingRules, userAttributes)) return false;
  return (await bucket(userId, flag.flagName, flag.seed)) < flag.rolloutPercent;
}